### PR TITLE
Except the pending banner things rest of the things has been done.

### PR DESCRIPTION
**Pull request by Azharhussain20**
1. Support new "Block" type: HTML.
2. Open app banner automatically when opening notification.
3. Support Font Family in Text Blocks

### CODE REVIEW

File: CleverPush/CleverPush/Source/CPAppBannerButtonBlock.m [hunk 1]
```diff
@@ -17,14 +17,16 @@ - (id)initWithJson:(NSDictionary*)json {
         } else {
             self.color = @"#000000";
         }
-        
+        if ([json objectForKey:@"family"]&& ![[json objectForKey:@"family"] isEqual:@""]) {
```
Comment (gbhrdt) [inline CleverPush/CleverPush/Source/CPAppBannerButtonBlock.m:20]: missing space before `&&`

File: CleverPush/CleverPush/Source/CPAppBannerController.m [hunk 2]
```diff
@@ -18,31 +18,31 @@ - (id)initWithBanner:(CPAppBanner*)banner {
...
         self.data = banner;
-        
+
         [self setModalPresentationStyle:UIModalPresentationCustom];
         [self setModalTransitionStyle:UIModalTransitionStyleCrossDissolve];
-        
+
         [self.view setContentMode:UIViewContentModeScaleToFill];
-        self.view.frame = CGRectMake(self.view.frame.origin.x, self.view.frame.origin.y, 414, 896);
+        self.view.frame = CGRectMake(self.view.frame.origin.x, self.view.frame.origin.y, [[UIScreen mainScreen] bounds].size.width, [[UIScreen mainScreen] bounds].size.height);
```
Comment (gbhrdt) [inline CleverPush/CleverPush/Source/CPAppBannerController.m:26]: what is that change for exactly?

File: CleverPush/CleverPush/Source/CPAppBannerController.m [hunk 3]
```diff
@@ -246,7 +251,12 @@ - (UILabel*)composeTextBlock:(CPAppBannerTextBlock*)block {
     
     label.text = block.text;
     label.textColor = [UIColor colorWithHexString:block.color];
-    [label setFont:[UIFont systemFontOfSize:(CGFloat)(block.size * 1.2) weight:UIFontWeightRegular]];
+    if ([CleverPush fontFamilyExits:block.family]){
+        [label setFont:[UIFont fontWithName:block.family size:(CGFloat)(block.size * 1.2)]];
+    }else{
```
Comment (gbhrdt) [inline CleverPush/CleverPush/Source/CPAppBannerController.m:256]: code style

File: CleverPush/CleverPush/Source/CPAppBannerController.m [hunk 4]
```diff
@@ -208,8 +208,13 @@ - (UIView*)composeButtonBlock:(CPAppBannerButtonBlock*)block {
     CPUIBlockButton *button = [CPUIBlockButton buttonWithType:UIButtonTypeSystem];
     [button setTitle:block.text forState:UIControlStateNormal];
     [button setTitleColor:[UIColor colorWithHexString:block.color] forState:UIControlStateNormal];
-    [button.titleLabel setFont:[UIFont systemFontOfSize:(CGFloat)(block.size * 1.2) weight:UIFontWeightSemibold]];
-    
+    if ([CleverPush fontFamilyExits:block.family]){
+        [button.titleLabel setFont:[UIFont fontWithName:block.family size:(CGFloat)(block.size * 1.2)]];
+    }else{
```
Comment (gbhrdt) [inline CleverPush/CleverPush/Source/CPAppBannerController.m:213]: code style

File: CleverPush/CleverPush/Source/CPAppBannerController.m [hunk 5]
```diff
@@ -293,6 +303,41 @@ - (UIImageView*)composeImageBlock:(CPAppBannerImageBlock*)block {
...
+        webView.configuration.defaultWebpagePreferences.allowsContentJavaScript = true;
+        webView.allowsBackForwardNavigationGestures = false;
+        webView.contentMode = UIViewContentModeScaleToFill;
+        webView.navigationDelegate = self;
+        webView.layer.cornerRadius = 15.0;
+        NSURL *url = [NSURL URLWithString:block.url];
+        NSURLRequest *request = [NSURLRequest requestWithURL:url];
+        [webView loadRequest:request];
+    }
+
```
Comment (gbhrdt) [inline CleverPush/CleverPush/Source/CPAppBannerController.m:324]: too much lines

File: CleverPush/CleverPush/Source/CPAppBannerController.m [hunk 6]
```diff
@@ -334,22 +379,25 @@ - (void)userContentController:(WKUserContentController*)userContentController
...
+                }
+                else
+                {
+                    CGFloat ScreenHeight = UIScreen.mainScreen.bounds.size.height;
+                    CGFloat popupY = (ScreenHeight - height) / 2;
+                    self.bannerBody.frame = CGRectMake(20, popupY , self.bannerBody.frame.size.width, height + 40);
+                    self.bannerBodyContent.frame = CGRectMake(20, 20, self.bannerBodyContent.frame.size.width, self.bannerBody.frame.size.height - 40);
+                    webView.frame = CGRectMake(0, 0, self.bannerBodyContent.frame.size.width, self.bannerBodyContent.frame.size.height);
+                }
+            }else{
```
Comment (gbhrdt) [inline CleverPush/CleverPush/Source/CPAppBannerController.m:399]: empty else block?

File: CleverPush/CleverPush/Source/CPAppBannerHTMLBlock.h [hunk 7]
```diff
@@ -0,0 +1,20 @@
+
```
Comment (gbhrdt) [inline CleverPush/CleverPush/Source/CPAppBannerHTMLBlock.h:1]: empty line

File: CleverPush/CleverPush/Source/CleverPush.m [hunk 8]
```diff
@@ -110,6 +110,7 @@ @implementation CleverPush
 CPHandleSubscribedBlock handleSubscribedInternal;
 DWAlertController *channelTopicsPicker;
 CPNotificationOpenedResult* pendingOpenedResult = nil;
+NSMutableArray<CPAppBanner*> *Pendingbanners;
```
Comment (gbhrdt) [inline CleverPush/CleverPush/Source/CleverPush.m:113]: variable naming change to: `pendingBanners`

File: CleverPush/CleverPush/Source/CleverPush.m [hunk 9]
```diff
@@ -2381,6 +2394,30 @@ + (CleverPush*)sharedInstance {
...
+    if (fontFamily == nil) {
+        return NO;
+    }
+
+    UIFontDescriptor *fontDescriptor = [UIFontDescriptor fontDescriptorWithFontAttributes:@{NSFontAttributeName:fontFamily}];
+    NSArray *matches = [fontDescriptor matchingFontDescriptorsWithMandatoryKeys: nil];
+
+    return ([matches count] > 0);
+}
++ (void)disableAppBanners{
```
Comment (gbhrdt) [inline CleverPush/CleverPush/Source/CleverPush.m:2408]: no new line

File: CleverPush/CleverPush/Source/CleverPush.m [hunk 10]
```diff
@@ -2381,6 +2394,30 @@ + (CleverPush*)sharedInstance {
     
     return singleInstance;
 }
++ (BOOL)fontFamilyExits:(NSString*)fontFamily
```
Comment (gbhrdt) [inline CleverPush/CleverPush/Source/CleverPush.m:2397]: no new line

**Review by gbhrdt — CHANGES_REQUESTED** [review]

File: CleverPush/CleverPush/Source/CPAppBannerController.m (hunk 2 above)
Comment (Azharhussain20) [reply CleverPush/CleverPush/Source/CPAppBannerController.m:26]: Actually previously it was a static frame height and width respectively 414, 896 so I have made them dynamic based on the size of device.

File: CleverPush/CleverPush/Source/CPAppBannerController.m (hunk 6 above)
Comment (Azharhussain20) [reply CleverPush/CleverPush/Source/CPAppBannerController.m:399]: actually there are two WKWebView in the same controller file to bifurcate the type of WKWebView on the delegate method it was developed before we havent decided about the height about the HTML block so it was forgotten to remove

**Pull request merged**